prow/gcs/reporter/reporter.go

```diff
@@ -0,0 +1,228 @@
+	passed := pj.Status.State == prowv1.SuccessState
+	f := metadata.Finished{
+		Timestamp: &completion,
+		Passed:    &passed,
+		Metadata:  metadata.Metadata{"uploader": "crier"},
+		Result:    string(pj.Status.State),
+	}
+	output, err := json.Marshal(f)
+	if err != nil {
+		return err
+	}
+
+	bucketName, dir, err := gr.getJobDestination(pj)
+	if err != nil {
+		return err
+	}
+
+	if gr.dryRun {
+		gr.logger.Infof("Would upload finished.json info to %q/%q", bucketName, dir)
+		return nil
+	}
+	b := gr.storage.Bucket(bucketName)
+	w := b.Object(path.Join(dir, "finished.json")).If(storage.Conditions{DoesNotExist: true}).NewWriter(ctx)
+	return gr.writeContent(w, output)
+}
+
+func (gr *gcsReporter) reportProwjob(ctx context.Context, pj *prowv1.ProwJob) error {
+	// Unconditionally dump the prowjob to GCS, on all job updates.
+	output, err := json.Marshal(pj)
+	if err != nil {
+		return err
+	}
+
+	bucketName, dir, err := gr.getJobDestination(pj)
+	if err != nil {
+		return err
+	}
+
+	if gr.dryRun {
+		gr.logger.Infof("Would upload pod info to %q/%q", bucketName, dir)
+		return nil
+	}
+	b := gr.storage.Bucket(bucketName)
+	w := b.Object(path.Join(dir, "prowjob.json")).NewWriter(ctx)
+	return gr.writeContent(w, output)
+}
+
+func (gr *gcsReporter) writeContent(w *storage.Writer, content []byte) error {
+	_, err := w.Write(content)
+	var reportErr error
+	if isErrUnexpected(err) {
+		reportErr = err
+		gr.logger.WithError(err).WithFields(logrus.Fields{"bucket": w.Bucket, "name": w.Name}).Warn("Uploading info to GCS failed (write)")
+	}
+	err = w.Close()
+	if isErrUnexpected(err) {
+		reportErr = err
+		gr.logger.WithError(err).WithFields(logrus.Fields{"bucket": w.Bucket, "name": w.Name}).Warn("Uploading info to GCS failed (close)")
+	}
+	return reportErr
+}
+
+func isErrUnexpected(err error) bool {
+	if err == nil {
+		return false
+	}
+	// Precondition Failed is expected and we can silently ignore it.
+	if e, ok := err.(*googleapi.Error); ok {
+		if e.Code == http.StatusPreconditionFailed {
+			return false
+		}
+		return true
+	}
+	return true
+}
+
+func (gr *gcsReporter) getJobDestination(pj *prowv1.ProwJob) (bucket string, dir string, err error) {
+	// The decoration config is always provided for decorated jobs, but many
+	// jobs are not decorated, so we guess that we should use the default location
+	// for those jobs. This assumption is usually (but not always) correct.
+	// The TestGrid configurator uses the same assumption.
+	ddc := gr.cfg().Plank.GetDefaultDecorationConfigs(pj.Spec.Refs.Repo)
+
+	var gcsConfig *prowv1.GCSConfiguration
+	if pj.Spec.DecorationConfig != nil && pj.Spec.DecorationConfig.GCSConfiguration != nil {
+		gcsConfig = pj.Spec.DecorationConfig.GCSConfiguration
+	} else if ddc != nil && ddc.GCSConfiguration != nil {
+		gcsConfig = ddc.GCSConfiguration
+	} else {
+		return "", "", fmt.Errorf("couldn't figure out a GCS config for %q", pj.Spec.Job)
+	}
+
+	_, dir, _ = gcsupload.PathsForJob(gcsConfig, &downwardapi.JobSpec{
```

Comment:
nit: Use https://github.com/kubernetes/test-infra/blob/ff6c355751962efd5d6b44f0f07567a9be1ea4f9/prow/pod-utils/downwardapi/jobspec.go#L49-L60

Reply:
Cool! I have wanted this many times and did not know it was there.

Reply:
Can you set this to something other than dir? If you're using dir as a return value, please only set it with a return statement, don't use naked returns in non trivial methods

https://github.com/golang/go/wiki/CodeReviewComments#named-result-parameters